File: algorithms/0042.trapping_rain_water/trapping_rain_water.cpp
class Solution {
public:
    int trap(const vector<int>& height) {
        int ans = 0;
        stack<size_t> s;
        for (size_t i = 0; i < height.size(); ++i) {
            while (!s.empty() && height[i] > height[s.top()]) {
                const size_t cur = s.top(); s.pop();
                if (s.empty()) {
                    break;
                }
                const size_t left = s.top();
                const size_t right = i;
                const size_t h = min(height[left], height[right]) - height[cur];
                ans += (right - left - 1) * h;
            }
            s.emplace(i);
        }
        return ans;
    }
};

File: algorithms/0042.trapping_rain_water/trapping_rain_water.rs
impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        let mut stack = Vec::<usize>::new();
        let mut ans = 0;
        for i in 0..height.len() {
            while !stack.is_empty() && height[*stack.last().unwrap()] < height[i] {
                let cur = *stack.last().unwrap();
                stack.pop();
                if stack.is_empty() {
                    break;
                }
                let left = *stack.last().unwrap();
                let right = i;
                let distance = height[right].min(height[left]) - height[cur];
                ans += distance * (right - left - 1) as i32;
            }
            stack.push(i);
        }
        ans
    }
}